src/buffer/out/TextAttribute.hpp

```diff
@@ -27,22 +27,38 @@ Revision History:
 #include "WexTestClass.h"
 #endif
 
+enum class UnderlineStyle : uint8_t
+{
+    NoUnderline = 0,
+    SinglyUnderlined = 1,
+    DoublyUnderlined = 2,
+    CurlyUnderlined = 3,
+    DottedUnderlined = 4,
+    DashedUnderlined = 5,
+};

```

Comment:
Instead of storing the underline style in a separate attribute, which ends up requiring an additional two bytes, I would much prefer if we could just add it to the existing `CharacterAttributes` enum.

I think all we really need is one more bit. Maybe move the `Faint` to 0x20, and then reserve 0x40, 0x80, and 0x100 for the underline styles. Something like this:

```cpp
SingleUnderline = 0x40,
DoubleUnderline = 0x80,
CurlyUnderline = 0xC0,
DottedUnderline = 0x100,
DashedUnderline = 0x140,
AnyUnderline = 0x1C0,
NoUnderline = 0,
```

Then you can access those attributes with something like this (pseudo code):

```cpp
// Testing for a particular style from the list above.
bool TextAttribute::IsUnderlined(int style)
{
    return (_attrs & AnyUnderline) == style;
}

// Testing for any underline style if we want a separate method for that.
bool TextAttribute::IsUnderlined()
{
    return IsUnderlined(NoUnderline);
}

// Setting a particular style from the list above.
void TextAttribute::SetUnderlined(int style)
{
    _attrs = (_attrs & ~AnyUnderline) | style;
}

// Remove all underlines if we want a separate method for that.
void TextAttribute::ResetUnderlined()
{
    SetUnderlined(NoUnderline);
}
```

There are probably some `WI_` macros you can use that'll make the implementation cleaner - I just wanted to demonstrate the general idea.

Reply:
Excellent! I think I can implement this, with little tweaks maybe.

Wondering if it's safe to assume only one of the underline style can be active at any point, including doublyUnderlined even if it's done using `SGR 21`. I guess we don't need to track doubly underline separately, do we?

Reply:
> I guess we don't need to track doubly underline separately, do we?

I'm glad you asked this, because we do track them separately at the moment, but I don't think we need to retain that behavior.

At the time I added double underline support there was no consensus in how this was handled amongst other terminals, so I just followed what XTerm was doing. But in retrospect I think that was probably the wrong decision, and now that we're adding more underline styles it makes even less sense. So yeah, I'd say we don't need to track double underline separately.

Reply:
I took a look at this discussion over at https://github.com/microsoft/terminal/issues/2916#issuecomment-652061927, and you mentioned:
<blockquote>
However in XTerm, underline and double-underline are actually separate attributes, and double-underline takes precedence when both are set, regardless of the order they were applied.
</blockquote>

I assume this needs to be taken care of?

Reply:
_SGR 4_ would check the current underline style, and only apply singleUnderline if the current style is not doublyUnderline.

This shouldn't apply to _SGR 4:x_.

Reply:
> > However in XTerm, underline and double-underline are actually separate attributes, and double-underline takes precedence when both are set, regardless of the order they were applied.
> 
> I assume this needs to be taken care of?

No. That's what I'm saying. It seemed like the right thing to do at the time, but in retrospect, copying XTerm was the wrong decision. I don't have a record of my testing from back then, but having just checked now, I can't find _anyone_ that still matches XTerm other than us. And in our case, we only really match XTerm in the DX renderer - the Atlas renderer draws the single underline on top of the double underline when both are applied.

There is no reason for us to retain the current behavior. This is such a ridiculous edge case anyway that nobody is going to be expecting it to work in a particular way.